src/thor/expansion_action.cc

```diff
@@ -0,0 +1,168 @@
+  float gen_factor = options.has_generalize() ? options.generalize() : 10.f;
+
+  // default the expansion geojson so its easy to add to as we go
+  Document dom;
+  dom.SetObject();
+  // set algorithm to Dijkstra, will be overwritten by other algos
+  SetValueByPointer(dom, "/properties/algorithm", "unidirectional_dijkstra");
+  SetValueByPointer(dom, "/type", "FeatureCollection");
+  SetValueByPointer(dom, "/features/0/type", "Feature");
+  SetValueByPointer(dom, "/features/0/geometry/type", "MultiLineString");
+  SetValueByPointer(dom, "/features/0/geometry/coordinates", Value(kArrayType));
+  SetValueByPointer(dom, "/features/0/properties", Value(kObjectType));
+  for (const auto& prop : options.expansion_props()) {
+    rapidjson::Pointer(kPropPaths[prop]).Set(dom, Value(kArrayType));
+    exp_props.insert(static_cast<Options_ExpansionProps>(prop));
+  }
+
+  // a lambda that the path algorithm can call to add stuff to the dom
+  // route and isochrone produce different GeoJSON properties
+  auto track_expansion = [&dom, &opp_edges, &gen_factor, &skip_opps,
+                          &exp_props](baldr::GraphReader& reader, baldr::GraphId edgeid,
+                                      const char* algorithm = nullptr, const char* status = nullptr,
+                                      const float duration = 0.f, const uint32_t distance = 0,
+                                      const float cost = 0.f) {
+    auto tile = reader.GetGraphTile(edgeid);
+    if (tile == nullptr) {
+      LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
+                std::to_string(edgeid.Tile_Base()));
+      return;
+    }
+    const auto* edge = tile->directededge(edgeid);
+    // unfortunately we have to call this before checking if we can skip
+    // else the tile could change underneath us when we get the opposing
+    auto shape = tile->edgeinfo(edge).shape();
+
+    // if requested, skip this edge in case its opposite edge has been added
+    // before (i.e. lower cost) else add this edge's id to the lookup container
+    if (skip_opps) {
+      auto opp_edgeid = reader.GetOpposingEdgeId(edgeid, tile);
+      if (opp_edgeid && opp_edges.count(opp_edgeid))
+        return;
+      opp_edges.insert(edgeid);
+    }
+
+    if (!edge->forward())
+      std::reverse(shape.begin(), shape.end());
+    Polyline2<PointLL>::Generalize(shape, gen_factor, {}, false);
+
+    // make the geom
+    auto& a = dom.GetAllocator();
+    auto* coords = GetValueByPointer(dom, "/features/0/geometry/coordinates");
+    coords->GetArray().PushBack(Value(kArrayType), a);
+    auto& linestring = (*coords)[coords->Size() - 1];
+    for (const auto& p : shape) {
+      linestring.GetArray().PushBack(Value(kArrayType), a);
+      auto point = linestring[linestring.Size() - 1].GetArray();
+      point.PushBack(p.first, a);
+      point.PushBack(p.second, a);
+    }
+
+    // no properties asked for, don't collect any
```

Comment:
so previously you couldn't select the properties, you just always got the hardcoded ones. i think we should retain that behavior if not set? otherwise if none were asked for why are we even running the request? maybe im missing something but if none are asked for we should have returned a 400 in loki at the beginning OR we should have defaulted to the ones that already existed before you improved the API

Reply:
you could just want the geometries for some maps? one property adds quite a lot of stuff to the response, I thought this was the most natural way to "support" that

Reply:
ok cool, so that is a good case for no additional properties other than geom. thanks for pointing it out. i guess i'll note that this is an API change along with getting full geom rather than just nodes in the graph. thats probably ok since the API was previously undocumented.